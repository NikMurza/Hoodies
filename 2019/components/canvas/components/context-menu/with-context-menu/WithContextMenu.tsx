import * as React from 'react';
import {
  ContextMenuItem,
  VisualEditorActions,
  VisualEditorComponentData,
  VisualEditorShapeType,
  VisualEditorContextMenuIdentity,
  VisualEditorPosition,
} from '@common/components/lib/types/visual-editor';
import Konva from 'konva';
import ContextMenu from '@common/components/lib/canvas/components/context-menu/ContextMenu';

export interface WithContextMenuProps {
  id: string;
  type: VisualEditorShapeType;
  isContextMenuShow: boolean;
  items: ContextMenuItem[];
  onAction<T extends object>(action: VisualEditorActions, data: T): void;
  onAddComponent: (data: VisualEditorComponentData) => void;
  onRemoveComponent: (id: string) => void;
}

export const typeContextMenuId = (id: string, type: VisualEditorShapeType): string =>
  `context_menu_${type}_${id}`;

export const WithContextMenu = <P extends object>(
  WrappedComponent: React.ComponentType<P>,
): React.FC<P & WithContextMenuProps> => (props: WithContextMenuProps): JSX.Element => {
  const { id, type, isContextMenuShow, items, onAction, onAddComponent, onRemoveComponent } = props;

  const contextMenuId = React.useMemo(() => typeContextMenuId(id, type), [id, type]);

  const [contextMenuPosition, setContextMenuPosition] = React.useState<VisualEditorPosition>({
    x: 0,
    y: 0,
  });

  const triggerAction = React.useCallback(
    (contextMenuId: string): void => {
      onAction<VisualEditorContextMenuIdentity>(VisualEditorActions.ContextMenu, {
        contextMenuId,
      });
    },
    [onAction],
  );

  const onContextMenu = React.useCallback(
    (e: Konva.KonvaEventObject<PointerEvent>): void => {
      e.evt.preventDefault();

      const { clientX: x, clientY: y } = e.evt;

      // запоминаем координаты где показывать меню
      setContextMenuPosition({ x, y });

      // Закрываем предыдущее меню, если есть. Генерим action показать контекстное меню
      triggerAction('');
      triggerAction(contextMenuId);
    },
    [triggerAction],
  );

  React.useEffect(() => {
    if (isContextMenuShow) {
      const { x, y } = contextMenuPosition;

      // показываем контектное меню
      const data: VisualEditorComponentData = {
        id: contextMenuId,
        component: (
          <ContextMenu id={id} key={contextMenuId} x={x} y={y} items={items} onAction={onAction} />
        ),
      };
      onAddComponent(data);
    } else {
      // прячем контекстное меню
      triggerAction('');
      onRemoveComponent(contextMenuId);
    }
  }, [isContextMenuShow]);

  return <WrappedComponent {...(props as P)} onContextMenu={onContextMenu} />;
};

export default WithContextMenu;
