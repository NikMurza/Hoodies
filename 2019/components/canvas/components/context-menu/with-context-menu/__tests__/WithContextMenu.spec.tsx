import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import {
  VisualEditorActions,
  ContextMenuItem,
  VisualEditorShapeType,
} from '@common/components/lib/types/visual-editor';
import { WithContextMenu, typeContextMenuId } from '../WithContextMenu';

const props = {
  id: 'id1',
  isContextMenuShow: true,
  items: [
    {
      icon: '',
      name: 'Rename',
      action: VisualEditorActions.Edit,
    },
    {
      icon: '',
      name: 'Delete',
      action: VisualEditorActions.Delete,
    },
  ],
  onAction: jest.fn,
  onAddComponent: jest.fn,
  onRemoveComponent: jest.fn,
  onContextMenu: jest.fn,
};

describe('WithContextMenu HOC', () => {
  it('renders correctly', () => {
    const onActionSpy = spyOn(props, 'onAction');

    const HOC = WithContextMenu(() => <div />);
    const wrapper = shallow(<HOC type={VisualEditorShapeType.Circle} {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.simulate('contextmenu', {
      evt: {
        preventDefault: jest.fn,
        offsetX: 100,
        offsetY: 100,
      },
    });
    expect(onActionSpy).toBeCalledWith(VisualEditorActions.ContextMenu, {
      contextMenuId: 'context_menu_circle_id1',
    });
  });

  it('renders correctly show', () => {
    const onActionSpy = spyOn(props, 'onAction');

    const HOC = WithContextMenu(() => <div />);
    const wrapper = shallow(
      <HOC type={VisualEditorShapeType.Circle} {...props} isContextMenuShow />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.simulate('contextmenu', {
      evt: {
        preventDefault: jest.fn,
        offsetX: 100,
        offsetY: 100,
      },
    });
    expect(onActionSpy).toBeCalledWith(VisualEditorActions.ContextMenu, {
      contextMenuId: 'context_menu_circle_id1',
    });
  });

  it('renders correctly hidden', () => {
    const onActionSpy = spyOn(props, 'onAction');

    const HOC = WithContextMenu(() => <div />);
    const wrapper = shallow(
      <HOC type={VisualEditorShapeType.Circle} {...props} isContextMenuShow={false} />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.simulate('contextmenu', {
      evt: {
        preventDefault: jest.fn,
        offsetX: 100,
        offsetY: 100,
      },
    });
    expect(onActionSpy).toBeCalledWith(VisualEditorActions.ContextMenu, {
      contextMenuId: 'context_menu_circle_id1',
    });
  });

  it('typeContextMenuId', () => {
    expect(typeContextMenuId('test', VisualEditorShapeType.Circle)).toEqual(
      'context_menu_circle_test',
    );
  });
});
