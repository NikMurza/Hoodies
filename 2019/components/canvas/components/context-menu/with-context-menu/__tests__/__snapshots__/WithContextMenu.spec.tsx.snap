// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithContextMenu HOC renders correctly 1`] = `
<Component
  id="id1"
  isContextMenuShow={true}
  items={
    Array [
      Object {
        "action": "Edit",
        "icon": "",
        "name": "Rename",
      },
      Object {
        "action": "Delete",
        "icon": "",
        "name": "Delete",
      },
    ]
  }
  onAction={[Function]}
  onAddComponent={[Function]}
  onContextMenu={[Function]}
  onRemoveComponent={[Function]}
  type="circle"
/>
`;

exports[`WithContextMenu HOC renders correctly hidden 1`] = `
<Component
  id="id1"
  isContextMenuShow={false}
  items={
    Array [
      Object {
        "action": "Edit",
        "icon": "",
        "name": "Rename",
      },
      Object {
        "action": "Delete",
        "icon": "",
        "name": "Delete",
      },
    ]
  }
  onAction={[Function]}
  onAddComponent={[Function]}
  onContextMenu={[Function]}
  onRemoveComponent={[Function]}
  type="circle"
/>
`;

exports[`WithContextMenu HOC renders correctly show 1`] = `
<Component
  id="id1"
  isContextMenuShow={true}
  items={
    Array [
      Object {
        "action": "Edit",
        "icon": "",
        "name": "Rename",
      },
      Object {
        "action": "Delete",
        "icon": "",
        "name": "Delete",
      },
    ]
  }
  onAction={[Function]}
  onAddComponent={[Function]}
  onContextMenu={[Function]}
  onRemoveComponent={[Function]}
  type="circle"
/>
`;
